[brandon-nam] ip

brandon-nam opened the pull request:
# ekuD 
> “Battle not with monsters, lest ye become a monster, and if you gaze into the abyss, the abyss gazes also into you.” - Friedrich Nietzsche [(source)](https://www.navigatelifetexas.org/en/blog/article/when-you-gaze-into-the-abyss#:~:text=%E2%80%9CBattle%20not%20with%20monsters%2C%20lest,abyss%20gazes%20also%20into%20you.%E2%80%9D&text=Nietzsche's%20quote%20above%20has%20become,what%20you%20are%20made%20of.)

If you are struggling with getting your tasks done on time, fret not. Though at times he is not as confident as Duke, ekuD is still there to help you. He is: 

- text-based
- easy to learn 

To use him all you need to do is: 

1. download it from [here](https://github.com/brandon-nam/ip),
2. go to the project directory on the command line 
3. run the java -jar duke.jar, 
4. read the description (still working on it!) and start using it! 

Easy right? And it is **FREE**! Here are list of things ekuD is capable of.. 

- [x] Managing tasks
- [x] Managing deadlines
- [x] Managing events 
- [x] Finding tasks
...And many more to come! 

If you are a java programmer, you can also use the program to practice Java. Here is the main function: 
```
 public static void main(String[] args) {
        Locale.setDefault(Locale.ENGLISH);
        String logo = " ____        _        \n"
                + "|  _ \\ _   _| | _____ \n"
                + "| | | | | | | |/ / _ \\\n"
                + "| |_| | |_| |   <  __/\n"
                + "|____/ \\__,_|_|\\_\\___|\n";
        System.out.println("Hello from\n" + logo);
        new Duke().greeting();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

namdohyun01 committed:
Level 0. Rename, Greet, Exit

namdohyun01 committed:
Level 1. Echo

namdohyun01 committed:
refactor: made functions that receive inputs, check for exit, and echoes.

namdohyun01 committed:
feat: Add and List texts

namdohyun01 committed:
feat: mark as done

namdohyun01 committed:
chore: new class Task added, and some new static functions

namdohyun01 committed:
feat: unmark added

namdohyun01 committed:
Added Deadline, Event, and Todo

namdohyun01 committed:
A-TextUiTesting

namdohyun01 committed:
A-TextUiTesting Scanner issue resolved

namdohyun01 committed:
A-TextUiTesting Scanner issue resolved

namdohyun01 committed:
Level-5 Handle Errors

namdohyun01 committed:
Level-6 Delete

namdohyun01 committed:
Duke.java: Add saveFile function

namdohyun01 committed:
Revert "Duke.java: Add saveFile function"

This reverts commit 40038f6095050f590614fd36ed6a94875faa0b6c.

namdohyun01 committed:
Duke.java: Add saveFile function

It is a good idea to save the tasks in the local disk, so that user can pull it out whenever they want.

saveFile function in making the program one step closer to the actual task scheduler.

namdohyun01 committed:
Teach the chatbot how to understand dates and times

The deadlines for the deadline tasks should be accepted in a date format.
Teach the chatbot how to understand certain formats of dates and times.

namdohyun01 committed:
Add date formatting functionality for deadline

Deadlines need actual dates, and should filter out random texts.

Minor change to accept only a designated date format.

namdohyun01 committed:
Use OOP to refactor and organize code

OOP is essential in keeping code clean.
Let's use OOP to keep it neat.
- Add Command, Parser, TextUi, TaskList classes.

namdohyun01 committed:
Add Storage class

Purpose of OOP is to organize code.
Make another class for Storage that stores the TaskList data on hard disk.

Khaleelur-Rahman commented on src/main/java/Duke.java:
```diff
@@ -1,38 +1,53 @@
 public class Duke {
-    public static Task[] taskList = new Task[100];
-    public static int taskListIndex = 0;
+    public static ArrayList<Task> taskList = new ArrayList<>();
```
The naming convention could be a bit better, eg, taskLists in the plural form :)

Khaleelur-Rahman commented on src/main/java/Task.java:
```diff
@@ -4,7 +4,7 @@ public class Task {
 
     public Task(String title) throws DukeException {
```
The naming of the description can be changed to maybe "description" so that it's easier to understand.

Khaleelur-Rahman commented on src/main/java/Duke.java:
```diff
@@ -1,38 +1,53 @@
     }
-    public static void markAsDone(int index) {
-        Task target = taskList[index - 1];
+    public static void markAsDone(int index) throws DukeIndexOutOfBoundsException{
```
Good job in handling the outOfBounds Exception :)

Khaleelur-Rahman reviewed (commented):
Overall good job in handling most of the edge cases. Good code structure too.

namdohyun01 committed:
Add checkstyle function in gradle

Checkstyle is a useful function to have in our project.
Configure checkstyle in build.gradle to make it work.

namdohyun01 committed:
Merge branch 'add-gradle-support'

namdohyun01 committed:
Edit packaging

Packaging is important in keeping our code clean.
Move brandon.chatbot to src/main/java directory.

namdohyun01 committed:
Add ParseTest class

Parsing incorrect arguments produces UnknownCommand class instance.
Add JUnit test to check if missing arguments lead to UnknownCommand.

namdohyun01 committed:
Create a JAR File

Package the app as an executable JAR file so that it can be distributed easily.

namdohyun01 committed:
Add JavaDoc comments to non-private classes/methods

Add JavaDoc comments to non-private classes/methods for better understanding of the code.

namdohyun01 committed:
Fix coding standard violation

Keeping the convention for coding standard is important for readable code.
Fix any coding standard violation in the code.

namdohyun01 committed:
Add find by title command

Finding tasks according to their titles can be helpful for users to filter specific relevant tasks.
Add FindCommand class and modify Parser class so that the program parses find command.

namdohyun01 committed:
Merge tag 'A-CodingStandard'

# Conflicts:
#	src/main/java/brandon/chatbot/Duke.java
#	src/main/java/brandon/chatbot/commands/Command.java
#	src/main/java/brandon/chatbot/commands/CommandResult.java
#	src/main/java/brandon/chatbot/parser/Parser.java
#	src/main/java/brandon/chatbot/tasks/Deadline.java

namdohyun01 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/brandon/chatbot/parser/Parser.java
#	src/main/java/brandon/chatbot/tasks/Event.java

namdohyun01 committed:
Configure build.gradle to add jar file

Configure build.gradle to add jar file.

namdohyun01 committed:
Fix checkstyle violations

Checkstyle is a useful gradle plugin to check any style convention violations.
Run checkstyle and fix any existing convention violation in the code.

namdohyun01 committed:
Add GUI

Graphic user interface makes the program more accessible to users.
Add GUI using JavaFX to make the program better.

namdohyun01 committed:
Add greeting message and HelpCommand class

Users may need to learn how to use the chatbot.
Add greeting message with the help of HelpCommand for better user experience.

namdohyun01 committed:
Edit JAR Configuration in build.gradle

JAR file lets the user play with the chatbot.
Let's edit JAR configuration in  build.gradle file so it creates JAR with JavaFX dependencies.

namdohyun01 committed:
Change ekuD image to a more reflective one

ekuD image does not really represent its character right now.
Change the image to something that is more representative.

namdohyun01 committed:
Merge tag 'A-CheckStyle'

# Conflicts:
#	src/main/java/brandon/chatbot/storage/Storage.java
#	src/main/java/brandon/chatbot/tasks/Event.java
#	src/main/java/brandon/chatbot/ui/TextUi.java

lunaflight commented on src/main/java/brandon/chatbot/storage/Storage.java:
```diff
@@ -0,0 +1,29 @@
+        this.path = path;
+    }
+    public void save(TaskList tasks) throws IOException {
+        try {
+            if (!Files.exists(path)) {
+                Files.createFile(path);
+            }
+            String data = "";
+            for (Task task : tasks.getList()) {
+                data += task.getStatus() + "\n";
+            }
+            Files.write(path, data.getBytes());
```
This try-catch block seems to only catch exceptions thrown by `Files.write`. We can reduce the amount of nesting by only having 1 line in the try-catch block.

lunaflight commented on src/main/java/brandon/chatbot/storage/Storage.java:
```diff
@@ -0,0 +1,29 @@
+                data += task.getStatus() + "\n";
+            }
+            Files.write(path, data.getBytes());
+        } catch (Exception e) {
```
Perhaps catching all `Exception`s regardless could be improved to catching only the type of `Exception` thrown by `Files.write`.

lunaflight reviewed (commented):
Looks good! Just a few nits :+1:

namdohyun01 committed:
Add assert tests for better code quality

Assert can be used in many places to gain insights of code behaviors and possible bugs caused by programmers.
Add assert tests in various places for better debugging.

namdohyun01 committed:
Add tagging ability to tasks

Tagging may come handy with additional features like searching tasks by their tags.
Add tagging for better user experience.

namdohyun01 committed:
Add FindTaskByTagCommand to find tasks by tag

Finding tasks by their tags is a useful feature to have.
Allow finding tasks with common tags by adding a FindTaskByTagCommand.

namdohyun01 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java

namdohyun01 committed:
Merge tag 'A-Assertions'

namdohyun01 committed:
Merge branch 'C-Tagging'

# Conflicts:
#	src/main/java/brandon/chatbot/Message.java
#	src/main/java/brandon/chatbot/commands/AddTodoCommand.java
#	src/main/java/brandon/chatbot/commands/Command.java
#	src/main/java/brandon/chatbot/commands/FindCommand.java
#	src/main/java/brandon/chatbot/commands/UnknownCommand.java
#	src/main/java/brandon/chatbot/commands/taskcommands/AddDeadlineCommand.java
#	src/main/java/brandon/chatbot/commands/taskcommands/AddEventCommand.java
#	src/main/java/brandon/chatbot/parser/Parser.java
#	src/main/java/brandon/chatbot/tasks/Task.java
#	src/main/java/brandon/chatbot/tasks/Todo.java

namdohyun01 committed:
Merge commit 'd475cdb6e22e43fb7cc54d8ca85c0c7b2dfbadb0'

namdohyun01 committed:
Make better GUI

A good GUI attracts more users and organizes information for better user experience.
Edit DialogBox to make it more aesthetic and user-friendly.

namdohyun01 committed:
Add emojis for detailed personality

Emojis can better represent ekuD's personality.
Add emojis in the messages to better capture ekuD's personality.

namdohyun01 committed:
Add More Error Handling

Each error requires unique error feedback to the user for the user to correctly make use of the chatbot.
Add special error feedbacks for each error to notice the user what went wrong.

namdohyun01 committed:
Merge tag 'A-MoreErrorHandling'

namdohyun01 committed:
Merge tag 'A-Personality'

# Conflicts:
#	src/main/java/brandon/chatbot/commands/generalcommands/UnknownCommand.java
#	src/main/java/brandon/chatbot/commands/taskcommands/AddDeadlineCommand.java
#	src/main/java/brandon/chatbot/commands/taskcommands/AddEventCommand.java
#	src/main/java/brandon/chatbot/commands/taskcommands/AddTodoCommand.java
#	src/main/java/brandon/chatbot/commands/taskcommands/FindCommand.java
#	src/main/java/brandon/chatbot/commands/taskcommands/FindTaskByTagCommand.java
#	src/main/java/brandon/chatbot/commands/taskcommands/ListCommand.java
#	src/main/java/brandon/chatbot/tasks/Deadline.java
#	src/main/java/brandon/chatbot/tasks/Event.java
#	src/main/java/brandon/chatbot/tasks/Task.java

brandon-nam committed:
Create gradle.yml

GitHub Actions will run the workflow (as per the .yml file) every time certain project events are triggered (e.g., when a PR is updated, or the master branch is updated).

brandon-nam committed:
Merge pull request #2 from brandon-nam/A-CI

Create gradle.yml

namdohyun01 committed:
Add JavaDoc to fit CheckStyle Requirements

CheckStyle requirements are to be followed for better readability.
Follow the CheckStyle requirements and add JavaDoc.

namdohyun01 committed:
Merge commit 'e63fe7ad08c40611b7da1faf9c04b9d75535edf5'

brandon-nam committed:
Update README.md

namdohyun01 committed:
Refactor from Duke to ekuD

brandon-nam committed:
Update README.md

Update README.md for a user guide.

brandon-nam committed:
Add screenshots to README.md

brandon-nam committed:
Update README.md

namdohyun01 committed:
Edit Feedbacks

Help requires a better listing of commands available, and task-related commands can show the current list of tasks to the user.
Edit feedbacks to make them more informative and helpful.

namdohyun01 committed:
Merge branch 'master' of https://github.com/brandon-nam/ip

brandon-nam committed:
Add Ui.png

namdohyun01 committed:
Merge branch 'master' of https://github.com/brandon-nam/ip

namdohyun01 committed:
Edit DukeIndexOutOfBoundsException Message

namdohyun01 committed:
Rename the ekuD class to Ekud

ekuD is does not follow class naming convention.
Rename ekuD to Ekud.

namdohyun01 committed:
Edit JAR File Name

brandon-nam committed:
Add Ui.png

brandon-nam committed:
Delete docs/Ui.png

brandon-nam committed:
Add Ui.png

brandon-nam committed:
Update README.md

namdohyun01 committed:
Add Load File Functionality

Loading task list from the previous run of chatbot could improve user experience.
Add loading file functionality by reading the output.txt file using buffer.

namdohyun01 committed:
Merge commit '564d6f7e7754a8e2cf1d02acaefb75922596e366'

namdohyun01 committed:
Fix FindCommand

FindCommand does not work properly due to the change in the Task class.
Fix FindCommand to make it work properly again.
